pkg/controller/machineset/machineset_controller.go

```diff
@@ -253,6 +262,58 @@ func (c *Controller) handleErr(err error, key interface{}) {
 	c.queue.Forget(key)
 }
 
+func (c *Controller) getMachineSet(key string) (*clusteroperator.MachineSet, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespace) == 0 || len(name) == 0 {
+		return nil, fmt.Errorf("invalid key %q: either namespace or name is missing", key)
+	}
+	return c.machineSetsLister.MachineSets(namespace).Get(name)
+}
+
+func (c *Controller) checkClusterRunning(syncFunc func(string) error) func(string) error {
+	return func(key string) error {
```

Comment:
Might be helpful to have some debug logs sprinkled through here for some easier visibility as to why a cluster isn't yet marked ready.